### DISCUSSION TITLE
allowed characters uiListBox

### DISCUSSION THREAD

**Discussion by mtarenskeen** [RomanKubiak/ctrlr#218] [Q&A]
I have experienced problems using certain characters in a uiListBox uiListBoxContent. In some of my panels I uses a uiListBox to display a list of patchnames. The characters that I found that give problems are:

ASCII 34 (")
ASCII 39 (')
ASCII 44 (,)
ASCII 61 (=)

I have worked around the problem by writing some LUA code to replace these characters with a ".". But maybe someone knows a better solution to allow displaying such characters in a uiListBox? I don't need complete Unicode support for this purpose, but would be happy enough with 7 bits ASCII support for ASCII characters 32 to 126.

**Comment by Ted-juh** [comment]
Do you use string.char(arg)? 

So (") would become string.char(34)

**Reply by mtarenskeen** [reply]
Yes I use that in some of my LUA code. But that's not the problem.

**Comment by dobo365** [comment]
I have the same issue since long. Using
```
sTemp = sTemp:gsub("'", " ")	-- Replace ' by a space as it gives issues in the listbox
sTemp = sTemp:gsub("\"", " ")	-- Replace " by a space as it gives issues in the listbox
```

Must be either a Juce issue or a Ctrlr issue.

**Reply by mtarenskeen** [reply]
![uiListBox-test](https://user-images.githubusercontent.com/3509304/112116555-b31afb80-8bba-11eb-8506-427ecde118c9.png)
I am having no problems using the `"\"` backslash in my uiListBox.

I have made a little testpanel with a uiListBox and used modulatorname "LBOX".
I attached it to the following LUA code:

```
function makeList()
	local m = panel:getComponent("LBOX")
	local s = ""
	for i = 32, 127 do
		if (i == 34) or (i == 39) or (i == 44) or (i == 61) then
			s = s .. string.format("ascii %03d is : [ %s ]\n", i, ".")
		else
			s = s .. string.format("ascii %03d is : [ %s ]\n", i, string.char(i))
		end
	end
	m:setProperty("uiListBoxContent", s, false)
end
```

**Comment by Ted-juh** [comment]
As far as I can tell, those characters seem to break the xml-parser. The uiListbox skips the lineFeed when you use one of those characters. Even when you add a character like ' in the XML directly by doing: &#39 it just shows a "square" and it still skips the linefeed (&#10). 

Ok, after a search on the old forums. This is what Atom explained:
Quotes are special characters you can’t just type in single quotes this will break the parser.

There is a class "CharPointer" that seems to deal with UTF_8 and ASCII but while there is some mention about CharPointer in the LCore.cpp file with .def("getAddress", &CharPointer_UTF8::getAddress).. I couldn't get that to work.

But more confusing is that the listbox uses "drawText" to output the text in the listbox. When you use drawText in a customComponent it shows those characters. Well, besides the " sign that breaks the string. And the \ needs to be doubled \\ to show up.
![Capture](https://user-images.githubusercontent.com/71118185/112362994-53654300-8cd5-11eb-89fe-97ca15d48df0.PNG)

The reason one does break the code and the other doesn't is, I think, because the uiListBox puts/ gets its string in/ from a StringArray and the latter uiCustomComponent drawText gets painted on the fly without the string being taken from the StringArray.

So you could try overriding with L&F or recreate the ListBox in a uiCustomComponent. But I guess your resolution to replace those characters is a lot less hassle.

**Reply by mtarenskeen** [reply]
Can you explain to me what you mean with "_try overriding with L&F_" ?

**Reply by Godlike-Productions** [reply]
The reason it breaks is because ', ", / and , do stuff in LUA, so they have to be escaped.

I came across a LUA method somewhere to replace those characters automagically when they are parsed from a UI component, but I can't for the life of me remember where I came across it, and I didn't end up using it.  Though I probably should check on my panel for exceptions with those characters as they may break stuff.

**Reply by mtarenskeen** [reply]
The problems arise also if I don't use LUA and simply create a uiListBox and enter the content manually in Panel Mode in the Ctrlr GUI. Slashes and backslashes work fine by the way. But ', ", &, and = give problems.

For now I can live with the limitation, and replace those characters with a dot or space.
But it is strange. These are very common characters, we are not talking about supporting Chinese, Hebrew, or Arabic characters here :-)

**Reply by Ted-juh** [reply]
> Can you explain to me what you mean with "_try overriding with L&F_" ?

It's not necessarily "overriding", it's more like you set LookAndFeel for certain parts of a component. The input in the uiListboxContent gets put in a StringArray (I think) and then the uiListbox uses drawText to show the content of the StringArray. And it's the conversion from UTF-8 to ASCII (or the other way round) that restricts you from using certain characters because they break some code.

My reasoning was that if you could set your own LookAndFeel and get your uiListboxContent from a table and then use drawText to fill the uiListbox, you would have a workaround. But I have to admit that's not going to be easy for a uiListbox. There are a lot of components like Slider, Combobox, Label that are very well documented on how to change the way they look. But I can't find much about uiListbox. I have to investigate further and get back to you.

**Comment by dnaldoog** [comment]
> I have experienced problems using certain characters in a uiListBox uiListBoxContent. In some of my panels I uses a uiListBox to display a list of patchnames. The characters that I found that give problems are:
> 
> ASCII 34 (") ASCII 39 (') ASCII 44 (,) ASCII 61 (=)
> 
> I have worked around the problem by writing some LUA code to replace these characters with a ".". But maybe someone knows a better solution to allow displaying such characters in a uiListBox? I don't need complete Unicode support for this purpose, but would be happy enough with 7 bits ASCII support for ASCII characters 32 to 126.

You should be able to fix this with __string.format("%q",)__.

<pre>
local m=MemoryBlock("42 61 63 6B 77 61 72 64 73 20 2A 36 30 27 73 20") -- Backwards *60's
-- add to combo content
local record=string.format("%q",m:toString())
panel:getComboComponent("myCombo"):setProperty("uiComboContent",record,false)
</pre>

It will surround everything with double quotes, but it works!

**Comment by mtarenskeen** [comment]
Thanks John,

Your solution using string.format("%q", s)  works nicely and does 
indeed solve my problem, but looks ugly :-)
I am considering using it. Good to have a choice.

MT

Op dinsdag 24 januari 2023 om 16:46:58 -0800 schreef John Goodland 
***@***.***>:

**Reply by dnaldoog** [reply]
Hi Martin,

I agree &#8211; looks a bit cluttered, but it does fix a common problem! 

I only just stumbled on it! I think it's unique to _lua_.

https://pgl.yoyo.org/luai/i/string.format

**Comment by mtarenskeen** [comment]
One small problem still occurs with patchsnames that have a doublequote 
in it.
A patche that I have named

	Say "Hi!"

will be displayed as

	"Say \"Hi!\""

which is ugly, but at least it won't break my uiListBox.

MT

Op woensdag 25 januari 2023 om 00:08:16 -0800 schreef John Goodland 
***@***.***>:

**Comment by dnaldoog** [comment]
Hi Martin,

In that case probably about the only thing I can think of is to substitute the " for a '!

<pre>
--
-- Called when a mouse is down on this component
--

myMethod = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local myList = ""
    local t = {'Hi there "everyone"', "60's are cool", "It's complex!", ' Say "Hi"'}
    for _, v in ipairs(t) do
        local parsed = string.gsub(v, '"', "'")
        myList = string.format("%s\n%q", myList, parsed)
    end

    panel:getComboComponent("c"):setProperty("uiComboContent", myList, false)
end

[escape text for combo_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/10498173/escape.text.for.combo_1_0.zip)

</pre>

**Comment by mtarenskeen** [comment]
Yep, that's what I did.


Op woensdag 25 januari 2023 om 02:01:25 -0800 schreef John Goodland 
***@***.***>:

**Comment by synthmutt** [comment]
I had "borrowed" code from another panel to replace problematic characters in variable s with a backtick before loading preset names into the bank listbox:

s = s:gsub('[="\\']', '`')

The backtick gets displayed in the listbox which shows all preset names in a bank of presets, but I'm preserving the original character in the saved preset data (in a panel variable). Sounds like I need to replace commas also.

**Comment by dobo365** [comment]
I'm using:

	sText = sText:gsub("\"", "")	-- Removes double quotes		
	sText = sText:gsub("'", "")	-- Removes single quotes

**Comment by dnaldoog** [comment]
This issue will probably be fixed in version 5.6.34

**Reply by mtarenskeen** [reply]
Great!
